### PR TITLE
ARROW-5113: [C++] Fix DoPut with dictionary arrays, add tests

### PR DESCRIPTION
**Pull request by pitrou**
Also fix an error where C++ Flight failed receiving 0-sized batches.

### CODE REVIEW

**Comment by pitrou** [comment]
@wesm @lihalite This makes Flight DoPut work with dictionaries. It shows that `ipc::MessageReader` is sufficient and there shouldn't be any explicit wrangling with dictionaries.

File: cpp/src/arrow/flight/flight-test.cc [hunk 1]
```diff
@@ -387,6 +445,32 @@ TEST_F(TestFlightClient, NoTimeout) {
...
+}
+
+TEST_F(TestDoPut, DoPutDicts) {
+  auto descr = FlightDescriptor::Path({"dicts"});
+  BatchVector batches;
+  auto dict_values = ArrayFromJSON(utf8(), "[\"foo\", \"bar\", \"quux\"]");
+  auto ty = dictionary(int8(), dict_values);
+  auto schema = arrow::schema({field("f1", ty)});
+  // Make several batches
+  for (const char* json : {"[1, 0, 1]", "[null]", "[null, 1]"}) {
```
Comment (pitrou) [inline cpp/src/arrow/flight/flight-test.cc]: As a side note, if one of these batches is empty (0-length), then I get an error in `RecordBatchStreamReader` ("Expected body in IPC message of type record batch").

File: python/pyarrow/tests/test_flight.py [hunk 2]
```diff
@@ -32,19 +32,40 @@
...
+            pa.DictionaryArray.from_arrays([2, 1], dict_values)]),
+    ]
+    return pa.Table.from_arrays(data, names=['some_dicts'])
+
+
 class ConstantFlightServer(flight.FlightServerBase):
     """A Flight server that always returns the same data.
 
     See ARROW-4796: this server implementation will segfault if Flight
     does not properly hold a reference to the Table object.
```
Comment (pitrou) [inline python/pyarrow/tests/test_flight.py]: I realize this may not be true anymore...

**Review by ghost — APPROVED** [review]

**Comment by pitrou** [comment]
Thanks @lihalite . Hopefully you don't get too many conflicts on your PRs :-)

**Pull request closed without merging**